src/dmangle.d

```diff
@@ -88,6 +94,9 @@ private immutable char[TMAX] mangleChar =
     //              X   // variadic T t...)
     //              Y   // variadic T t,...)
     //              Z   // not variadic, end of parameters
+    //              %   // Type backward reference
+    //              ?   // symbol backward reference
+    //              $   // identifier backward reference
```

Comment:
> Maybe compression should be done higher up the grammar then? Whole types, whole identifier names.

Types are the main point of this PR, single identifier compression is actually an additional micro optimization. It's a special case of the symbol compression that could help with recursive nested function templates.

> Maybe I shouldn't point at C++ as a way of example, but their way of compressing symbols seems to be good enough for them (and unquestionably portable). We could use a system of Q_, Q1_, etc...

GCC's compression seems quite good, it's actually more or less what's copied for types here. VC's mangling performs similar bad as current D for large recursive template symbols because it can only back reference the first 10 types.